### PR TITLE
[examples] prototype of frenemies game

### PR DESCRIPTION
**Pull request by sblackshear**

### CODE REVIEW

File: crates/sui-framework/sources/tx_context.move [hunk 1]
```diff
@@ -88,10 +88,10 @@ module sui::tx_context {
     fun dummy_tx_hash_with_hint(hint: u64): vector<u8> {
         let tx_hash = vector[];
         let i = 1;
-        let hash_length = (TX_HASH_LENGTH as u8);
+        let hash_length = TX_HASH_LENGTH;
         while (i <= hash_length) {
-            let value = if (i <= 8) ((hint >> (64 - (8 * i))) as u8) else 0;
-            vector::push_back(&mut tx_hash, value);
+            let value = i ^ hint % 256;
```
Comment (emmazzz) [inline crates/sui-framework/sources/tx_context.move:93]: Not a big deal since it's test only but why this change?

File: crates/sui-framework/sources/math.move [hunk 2]
```diff
@@ -22,6 +22,15 @@ module sui::math {
         }
     }
 
+    /// Return the absolute value of x - y
+    public fun abs(x: u64, y: u64): u64 {
```
Comment (emmazzz) [inline crates/sui-framework/sources/math.move:26]: "abs" normally refers to the absolute value of one number. Maybe call this "distance" instead?

File: crates/sui-framework/sources/tx_context.move (hunk 1 above)
Comment (Ahmedpooh) [reply crates/sui-framework/sources/tx_context.move:93]: I don't know, can you explain this change?

File: crates/sui-framework/sources/math.move (hunk 2 above)
Comment (Ahmedpooh) [reply crates/sui-framework/sources/math.move:26]: Yrs interesting

File: sui_programmability/examples/games/sources/frenemies.move [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        scorecard: &mut Scorecard, state: &SuiSystemState, leaderboard: &mut Leaderboard, ctx: &mut TxContext
+    ) {
+        let cur_epoch = tx_context::epoch(ctx);
+        if (cur_epoch > leaderboard.epoch) {
+            // if first call in new epoch, update leaderboard with stake rankings for the previous epoch.
+            // Note: there's a pragmatically safe, but theoretically unsafe assumption here that
+            // `cur_epoch == leaderboard.epoch + 1` because at least one player will update their scorecard
+            // each epoch. if this assumption is violated via 0 calls to `update` during epoch `N`, players
+            // will not get credit for scores during epoch `N - 1`.
+            update_epoch_info(state, leaderboard, cur_epoch - 1)
```
Comment (emmazzz) [inline sui_programmability/examples/games/sources/frenemies.move:148]: I think it makes more sense to use `cur_epoch` here and maybe `cur_epoch + 1` for the `scorecard`'s epoch below. We should use `cur_epoch` here because the stake distribution stored in the system state is the stake distribution that's used in our system throughout the current epoch so using the current epoch number would be more intuitive.

File: sui_programmability/examples/games/sources/frenemies.move [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    use sui::tx_context::{Self, TxContext};
+    use sui::validator_set;
+    use sui::validator;
+
+    #[test_only]
+    use sui::test_scenario::{Self as ts, Scenario};
+
+    /// Singleton struct scoring staking infomration for recent epochs,
+    /// start epoch, and a leaderborad with top scorers
+    struct Leaderboard has key {
```
Comment (emmazzz) [inline sui_programmability/examples/games/sources/frenemies.move:19]: Not sure if "Leaderboard" is a good name. A leaderboard should contain players with highest scores, who are, in our case, players who have completed their goals and with highest difficulties. This doesn't really have that info right now.

File: sui_programmability/examples/games/sources/frenemies.move [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+    /// Get the current assignment for `addr`.
+    /// Always returns the same assignment for a fixed address A and epoch E.
+    fun get_assignment(addr: address, state: &SuiSystemState): Assignment {
+        let validators = validator_set::active_validators(sui_system::validators(state));
+        let len = vector::length(validators);
+
+        // derive assignment from the stake of top validator + address.
+        // this means that for a fixed address A and epoch E, the assignment is deterministic
+        let top_stake = (validator::total_stake(vector::borrow(validators, 0)) as u256);
```
Comment (emmazzz) [inline sui_programmability/examples/games/sources/frenemies.move:107]: Right now we just take the address of the first validator in the active validator vector, which is unsorted, so I wouldn't call it "top_stake".

File: sui_programmability/examples/games/sources/frenemies.move [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+            },
+            sender
+        )
+    }
+
+    /// Update `scorecard` with the results from the last epoch.
+    /// Note: relies on someone calling this at least once/epoch to preserve
+    /// the record of validator scores
+    public entry fun update(
+        scorecard: &mut Scorecard, state: &SuiSystemState, leaderboard: &mut Leaderboard, ctx: &mut TxContext
```
Comment (emmazzz) [inline sui_programmability/examples/games/sources/frenemies.move:139]: The leaderboard really just needs to be updated once per epoch and the system state object is solely here for that purpose. (It's also used right now for getting the assignment but we can use the stake info stored in the leaderboard instead.)

I think in general it's better to reduce the use of the shared system state object since it's also used in all the staking/delegation/tallying/gas price operations. Maybe we can have another version of the entry funs `register` and `update` that don't take in the system state obj and `assert` that `leaderboard.epoch == cur_epoch`.

**Review by emmazzz — DISMISSED** [review]
Left some comments about naming and the usage of the system state object.

File: crates/sui-framework/sources/tx_context.move (hunk 1 above)
Comment (sblackshear) [reply crates/sui-framework/sources/tx_context.move:93]: This was crashing with an arithmetic error--I think `hint` can be > 256 and the cast fails. I don't really understand the previous math, but I think all this code really needs to do is produce a different tx hash for each `hint` (which this fix will accomplish).

File: sui_programmability/examples/games/sources/frenemies.move (hunk 4 above)
Comment (sblackshear) [reply sui_programmability/examples/games/sources/frenemies.move:19]: Good point. I started off thinking about keeping the score in here, but wasn't sure whether a "top 100 or so" in a single field impl, a dynamic fields-based impl, or events-based leaderboard impl would be best from a frontend perspective.

I'll go ahead with a simple "top 500 in a single field" approach for now, but we can tweak this if needed.

File: sui_programmability/examples/games/sources/frenemies.move (hunk 6 above)
Comment (sblackshear) [reply sui_programmability/examples/games/sources/frenemies.move:139]: This broadly makes sense to me, but I'm worried about what will happen at the beginning of a new epoch. My expectation is that a lot of users will call `update` ASAP, but this will fail unless someone has called the function that adds the stake info to the `Leaderboard` already.

If we are willing to tweak the wave 2 epoch change code to do this, that would be one way to solve the problem, but I know we (reasonably) wanted the epoch change code to be as separate as possible from the game.

File: sui_programmability/examples/games/sources/frenemies.move (hunk 5 above)
Comment (sblackshear) [reply sui_programmability/examples/games/sources/frenemies.move:107]: Good point + will fix.

File: sui_programmability/examples/games/sources/frenemies.move (hunk 6 above)
Comment (emmazzz) [reply sui_programmability/examples/games/sources/frenemies.move:139]: Fair point. I also can't think of a good way of doing this without tweaking the advance epoch code, which I don't think we should do. If it's not too hard maybe we can add frenemies game operations to stress testing too, emitting them together with staking operations to see how we handle these.

File: crates/sui-framework/src/lib.rs [hunk 7]
```diff
@@ -152,13 +152,13 @@ pub fn run_move_unit_tests(
 /// Wrapper of the build command that verifies the framework version. Should eventually be removed once we can
 /// do this in the obvious way (via version checks)
 pub fn build_move_package(path: &Path, config: BuildConfig) -> SuiResult<CompiledPackage> {
-    let test_mode = config.config.test_mode;
+    //let test_mode = config.config.test_mode;
     let pkg = config.build(path.to_path_buf())?;
-    if test_mode {
+    /*if test_mode {
```
Comment (emmazzz) [inline crates/sui-framework/src/lib.rs:157]: Uncomment?

File: sui_programmability/examples/frenemies/sources/frenemies.move [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+        vector::push_back(v, validator)
+    }
+
+     /// write `s` into `v`, maintaining the invariant that `v` is in descending order by (score, participation)
+    fun score_insertion_sort(v: &mut vector<Score>, new_score: Score) {
+        let i = vector::length(v);
+        let score = new_score.score;
+        while (i > 0) {
+            let s = vector::borrow_mut(v, i - 1);
+            if (score > s.score || score == s.score && new_score.participation > s.participation) {
```
Comment (emmazzz) [inline sui_programmability/examples/frenemies/sources/frenemies.move:302]: This should be `new_score.participation > s.participation` right? Since less participation means higher rank

File: crates/sui-framework/src/lib.rs (hunk 7 above)
Comment (sblackshear) [reply crates/sui-framework/src/lib.rs:157]: Good call--I just did this because it's impossible to build/test w/ framework changes otherwise :(

File: sui_programmability/examples/frenemies/sources/frenemies.move (hunk 8 above)
Comment (sblackshear) [reply sui_programmability/examples/frenemies/sources/frenemies.move]: Yes, good catch! Fixed + tested much better in latest.

File: sui_programmability/examples/frenemies/sources/registry.move [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        })
+    }
+
+    /// Only callable from scorecard
+    public(friend) fun register(self: &mut Registry, name: String, ctx: &TxContext): Name {
+        assert!(string::length(&name) <= MAX_NAME_SIZE, ENameTooLong);
+
+        let name = Name { name };
+        let players = &mut self.players;
+        assert!(!table::contains(players, name), ENameAlreadyRegistered);
```
Comment (emmazzz) [inline sui_programmability/examples/frenemies/sources/registry.move:44]: With this setup, an address can register again but with a different name. Is this something we want to allow?

Comment (emmazzz) [reply sui_programmability/examples/frenemies/sources/registry.move:44]: Oh from the tests it seems that we are fine with it. Why though?

Comment (sblackshear) [reply sui_programmability/examples/frenemies/sources/registry.move:44]: There's no harm in registering multiple times with the same address, and it doesn't give a player any advantages because all the scorecards belonging to the same address will get the same assignment. We could prevent this with an extra table, but I thought I'd allow it so folks can explore the strategy of getting multiple scorecards and then discover that it doesn't help :)

**Review by emmazzz — APPROVED** [review]
